File: fluorophores/FlMoving.m
classdef FlMoving<Fluorophore
    properties
        % pos=[0, 0, 0]; %nm     
        posind=1;
        posparameters=[];
        posmode
    end
    methods
        % function reset(obj)
        %     reset@MFfluorophore(obj);
        %     %later: also reset trace / position if wanted
        % end
        function [posout,isactive]=position(obj,time,props)
            isactive=true;
            pos=obj.pos;
            posout=[0 0 0];
            posmode=obj.posmode;
            switch posmode
                case 'function'
                    for k=length(pos):-1:1
                        posh(k)=pos{k}(time);
                    end
                case 'static'
                    posh=pos;
                case {'trace','diffusion','steps'}
                    %XXXX add test if ind. already too high (e.g. 2x
                    %position asked. If asked 2x same time: should work)
                    posind=max(1,obj.posind-1); 
                    if pos(posind,1)>time
                        posind=find(pos(:,1)<time,1,'last');
                        if isempty(posind)
                            posind=1;
                        end
                    end
                    while pos(posind,1)<time
                        if posind>=size(pos,1)-1 %reached end
                            obj.extendtrace;
                            pos=obj.pos;
                        end
                        posind=posind+1;
                    end
                posh=pos(posind,2:end);
                obj.posind=posind;
            end
            posout(1:length(posh))=posh;
        end
        function makediffusion(obj,D,dt,args)
            arguments
                obj
                D %um^2/s
                dt %us
                args.startpos=[0 0 0]
                args.dim=2;
                args.numpoints=10000;
                args.boundarybox=[]; %nm, half length of box, periodic boundary conditions
            end
            obj.posparameters={D,dt,args};
            % Dstep=D/1e6*1e6*dt; %D in um^2/s is the same as D in nm^2/us.
            % to obtain nm^2/ms multiply by 1000;
            Dstep=D*dt*1000;
            time=dt:dt:args.numpoints*dt;
            jumps=(randn(args.numpoints,args.dim)*Dstep);
            pos=horzcat(cumsum(jumps,1))+args.startpos(1:args.dim);
            if ~isempty(args.boundarybox) %periodic boundary conditions
                if numel(args.boundarybox)==1
                    args.boundarybox=args.boundarybox.*ones(1,args.dim);
                end
                pos2=pos+args.boundarybox;
                pos3=mod(pos2,2*args.boundarybox);
                pos=pos3-args.boundarybox;
            end

            obj.pos=horzcat(time',pos);
            obj.posmode='diffusion';
        end
        function makesteps(obj,stepsize,dwelltime,dt,args)
            arguments
                obj
                stepsize %nm
                dwelltime %ms;
                dt %ms
                args.startpos=[0 0 0]
                args.dim=2;
                args.numpoints=1000;
                args.angle=0;
            end
            obj.posparameters={stepsize,dwelltime,dt,args};
            time=(dt:dt:args.numpoints*dt)';
            
            numjumps=ceil(args.numpoints*dt/(dwelltime)*2+5);
            jmp=exprnd(dwelltime/dt,numjumps,1);
            jumppos=(cumsum(jmp));
            jumppos(jumppos>length(time))=[];
            % xj=zeros(size(time));
            % xj(jumppos)=stepsize;
            xjh=histcounts(jumppos,(0:length(time)))*stepsize;
            xpos=cumsum(xjh');

            R = [cosd(args.angle) -sind(args.angle); sind(args.angle) cosd(args.angle)];
            pos=(R*horzcat(xpos,zeros(size(xpos)))')'+args.startpos(1:args.dim);
            obj.pos=horzcat(time,pos);
            obj.posmode='steps';
        end
        function extendtrace(obj)
            currentpos=obj.pos(end,:);
            oldpos=obj.pos;
            % obj.posparameters(3).startpos=currentpos(2:end);
            args=obj.posparameters{end};
            switch obj.posmode
                case 'diffusion'
                    obj.makediffusion(obj.posparameters{1},obj.posparameters{2},dim=args.dim,numpoints=args.numpoints,startpos=currentpos(2:end),boundarybox=args.boundarybox)
                case 'steps'
                    obj.makesteps(obj.posparameters{1},obj.posparameters{2},obj.posparameters{3},dim=args.dim,numpoints=args.numpoints,startpos=currentpos(2:end), angle=args.angle)
            end        
            obj.pos(:,1)=obj.pos(:,1)+currentpos(1); %time continues
            obj.pos=vertcat(oldpos,obj.pos);
        end
    end
end

File: python/fluorophores/FlMoving.py
import numpy as np

from .Fluorophore import Fluorophore

class FlMoving(Fluorophore):
    def __init__(self, pos=[0,0,0], brightness=1000):
        super().__init__(pos, brightness)
        self.posind = 0
        self.posparameters = []
        self.posmode = None
        self.posfunction = None

    def position(self, time, props=None):
        posout = np.array([0,0,0])
        posmode = self.posmode
        if posmode == "static":
            posh = self.pos
        elif posmode in ["trace", "diffusion", "steps"]:
            # XXXX add test if ind. already too high (e.g. 2x
            # position asked. If asked 2x same time: should work)
            pos = self.pos
            posind = np.maximum(0, self.posind-1)
            if pos[posind,0] > time:
                result = np.where(pos[:,0]<time)[0]
                if len(result) > 0:
                    posind = result[-1]
                else:
                    posind = 0
            while pos[posind,0] < time:
                if posind >= pos.shape[0]-1:  # reached end
                    self.extendtrace()
                    pos = self.pos
                posind += 1
            posh = pos[posind,1:]
            self.posind = posind
        elif posmode == "function":
             posfunction = self.posfunction
             for k in range(len(posfunction),-1,-1):
                  posh[k] = posfunction[k](time)
        else:
             raise ValueError(f"Posmode {posmode} not implemented.")
        
        posout[:len(posh)] = posh

        return posout, True
    
    def makediffusion(self, 
                      D, # um^2/s
                      dt,  # us
                      startpos=np.array([0,0,0]), 
                      dim=2,  # number of dimensions to simulate
                      numpoints=10000, 
                      boundarybox=None): # nm, half length of box, periodic boundary conditions
        # print("makediffusion")
        kwargs = {'startpos': startpos, 
                  'dim': dim, 
                  'numpoints': numpoints, 
                  'boundarybox': boundarybox}
        self.posparameters = [D, dt, kwargs]

        # to obtain nm^2/ms multiply by 1000;
        Dstep = D*dt*1000
        time = np.atleast_2d((np.arange(numpoints)+1)*dt)
        jumps = np.random.randn(numpoints, dim)*Dstep
        # print("max jump: ", jumps.max(0))
        # print(f"startpos: {startpos[:dim]}")
        pos = np.cumsum(jumps, axis=0) + startpos[:dim]
        # print("max pos: ", pos.max(0))
        if boundarybox is not None:  # periodic boundary conditions
            if not (isinstance(boundarybox, list) or isinstance(boundarybox, np.ndarray)) or len(boundarybox) == 1:
                boundarybox = boundarybox*np.ones((1,dim))
            if not isinstance(boundarybox, np.ndarray):
                boundarybox = np.array(boundarybox)
            boundaryboxpos = boundarybox/2
            pos2 = pos + boundaryboxpos
            pos3 = np.mod(pos2, 2*boundaryboxpos)
            pos = pos3 - boundaryboxpos
        # print("max pos after boundary box: ", pos.max(0))
        
        self.pos = np.hstack([time.T, pos])
        self.posmode = "diffusion"

    def makesteps(self, 
                  stepsize,  # nm
                  dwelltime,  # ms
                  dt, 
                  startpos=np.array([0,0,0]), 
                  dim=1, 
                  numpoints=10000,
                  angle=0):
        # print("makesteps")
        kwargs = {'startpos': startpos, 
                  'dim': dim, 
                  'numpoints': numpoints, 
                  'angle': angle}
        self.posparameters = [stepsize, dwelltime, dt, kwargs]

        time = np.atleast_2d(((np.arange(numpoints)+1)*dt)).T

        numjumps = np.ceil(numpoints*dt/(dwelltime)*2+5).astype(int)
        jmp=np.random.exponential(dwelltime/dt,(numjumps,1))
        jumppos=np.cumsum(jmp)
        jumppos = jumppos[jumppos<=len(time)]
        hist, bin_edges = np.histogram(jumppos, bins=np.arange(len(time) + 1))
        xjh = hist * stepsize
        xpos=np.cumsum(xjh.T)
        angle_rad = angle*np.pi/180
        
        R = np.array([[np.cos(angle_rad) - np.sin(angle_rad)], 
                      [np.sin(angle_rad), np.cos(angle_rad)]])
        pos = (R*np.hstack([xpos, np.zeros(xpos.shape)]).T).T + startpos[:dim]
        self.pos = np.hstack([time, pos])
        self.posmode = "steps"

    def extendtrace(self):
        currentpos = self.pos[-1,:]
        oldpos = self.pos.copy()
        kwargs = self.posparameters[-1]
        if self.posmode == "diffusion":
            self.makediffusion(self.posparameters[0], 
                               self.posparameters[1],
                               dim=kwargs['dim'],
                               numpoints=kwargs['numpoints'],
                               startpos=currentpos[1:],
                               boundarybox=kwargs['boundarybox'])
        elif self.posmode == "steps":
            self.makesteps(self.posparameters[0],
                           self.posparameters[1],
                           self.posparameters[2],
                           dim=kwargs['dim'],
                           numpoints=kwargs['numpoints'],
                           startpos=currentpos[1:],
                           angle=kwargs['angle'])
        
        self.pos[:,0] += currentpos[0]  # time continues
        self.pos = np.vstack([oldpos, self.pos])
